### PR TITLE
Add GUC to control join qual propagation

### PR DESCRIPTION
**Pull request by svenklemm**
This patch adds a `enable_qual_propagation` GUC to control
propagation of JOIN quals. Since there have been a few instances
where JOIN qual propagation was too aggressive this GUC can be
used to disable qual propagation.

### CODE REVIEW

File: src/plan_expand_hypertable.c [hunk 1]
```diff
@@ -1145,7 +1145,7 @@ ts_plan_expand_timebucket_annotate(PlannerInfo *root, RelOptInfo *rel)
 	/* Walk the tree and find restrictions or chunk exclusion functions */
 	timebucket_annotate_walker((Node *) root->parse->jointree, &ctx);
 
-	if (ctx.propagate_conditions != NIL)
+	if (ctx.propagate_conditions != NIL && ts_guc_enable_qual_propagation)
```
Comment (mkindahl) [inline src/plan_expand_hypertable.c:1148]: It is easier to maintain this code if the check is put inside the `propagate_join_quals`. That way, if a call is added somewhere else, there is no risk that a join propagation is done for that case, which means easier debugging.

**Review by mkindahl — COMMENTED** [review]
Looks good, but I suggest to move the check to inside `propagate_join_quals` instead since this is easier to maintain and reduced the risk of introducing subtle bugs.

**Review by gayyappan — APPROVED** [review]

**Comment by erimatnor** [comment]
A couple of thoughts on this:

* I am not sure users would intuitively understand what it means to set `enable_qual_propagation=false` (or `true`). This seems way too implementation specific. What's a "qual"? Where are they propagated? What does having it on or off mean?
* I think we should be very restrictive with introducing new GUCs. They're essentially APIs that require documentation and commitments going forward.
* IMO, we shouldn't have GUCs to disable improper behavior. If something is not working correctly (even if only in some cases) we shouldn't do it until we can fix the issue. 
* When we fix the underlying issue (i.e., so that quals aren't propagated too aggressively), then what happens to this GUC?

**Comment by mkindahl** [comment]
> A couple of thoughts on this:
> 
>     * I am not sure users would intuitively understand what it means to set `enable_qual_propagation=false` (or `true`). This seems way too implementation specific. What's a "qual"? Where are they propagated? What does having it on or off mean?

IMHO, these are intended for debugging and implementing workarounds for cases that do not work. They are a last resort for users when things break.

> 
>     * I think we should be very restrictive with introducing new GUCs. They're essentially APIs that require documentation and commitments going forward.

These are debugging and last-resort knobs and as such I do think there is a point that it is good to document them, but I do not think that is a strict requirement.

> 
>     * IMO, we shouldn't have GUCs to disable improper behavior. If something is not working correctly (even if only in some cases) we shouldn't do it until we can fix the issue.

We will always risk that some feature do not work, however well we test it. This knob is not introduced to circumvent a known issue (a fix is already in the pipe), this is to be able to disable a behavior that we have discovered generate problems so that it can be disabled if we discover yet another issue. 

> 
>     * When we fix the underlying issue (i.e., so that quals aren't propagated too aggressively), then what happens to this GUC?

When it is not needed any more, it will be removed.

In summary: this is a debugging and workaround tool to be able to turn off functionality that we discover is unstable. It is not an alternative to fixing known issues: it is intended as a last resort when you hit a not previously discovered issue.

**Review by mkindahl — APPROVED** [review]

**Review by k-rus — COMMENTED** [review]
We already have GUC to enable or disable Timescale optimisation (`ts_guc_enable_optimizations`). Why cannot it be used for this?
Otherwise, should each optimisation implemented in TimescaleDB introduce new GUC?

**Comment by svenklemm** [comment]
> We already have GUC to enable or disable Timescale optimisation (`ts_guc_enable_optimizations`). Why cannot it be used for this?
> Otherwise, should each optimisation implemented in TimescaleDB introduce new GUC?

enable_optimizations is not really a practical workaround as it disables everything, and is not really practical to run timescaledb with this flag set. Setting this flag also doesn't give you additional information where behaviour is originating from. Having flags for specific optimizations is very useful as you can run queries with certain optimizations enabled/disabled to pinpoint the source of the behaviour. Also running with join qual propagation disabled shouldn't have big impact on normal operation and you can achieve similar results by rewriting queries.

This GUC is not meant as public API but as a tool for support.
